packages/protocol/scripts/truffle/make-release.ts

```diff
@@ -0,0 +1,219 @@
+        } else {
+          this.dependencies.set(contract, [lib])
+        }
+      })
+    })
+  }
+
+  public get = (contract: string): string[] => {
+    return this.dependencies.has(contract) ? this.dependencies.get(contract) : []
+  }
+}
+
+class ContractAddresses {
+  constructor(public addresses: Map<string, Address>) {}
+  static async create(contracts: string[], registry: any) {
+    const addresses = new Map()
+    contracts.forEach(async (contract: string) => {
+      const registeredAddress = await registry.getAddressForString(contract)
+      if (!eqAddress(registeredAddress, NULL_ADDRESS)) {
+        addresses.set(contract, registeredAddress)
+      }
+    })
+    return new ContractAddresses(addresses)
+  }
+
+  public get = (contract: string): Address => {
+    if (this.addresses.has(contract)) {
+      return this.addresses.get(contract)
+    } else {
+      throw new Error(`Unable to find address for ${contract}`)
+    }
+  }
+
+  public set = (contract: string, address: Address) => {
+    this.addresses.set(contract, address)
+  }
+}
+
+module.exports = async (callback: (error?: any) => number) => {
+  try {
+    const argv = require('minimist')(process.argv.slice(2), {
+      string: ['report', 'network', 'proposal', 'libraries', 'initialize_data', 'build_directory'],
+    })
+    const report: ASTDetailedVersionedReport = readJsonSync(argv.report).report
+    const initializationData = readJsonSync(argv.initialize_data)
+    const dependencies = new ContractDependencies(linkedLibraries)
+    const contracts = readdirSync(join(argv.build_directory, 'contracts')).map((x) =>
+      basename(x, '.json')
+    )
+    const registry = await artifacts
+      .require('Registry')
+      .at('0x000000000000000000000000000000000000ce10')
+    const addresses = await ContractAddresses.create(contracts, registry)
+    const released: Set<string> = new Set([])
+    const proposal = []
+    // Release 1 will deploy all libraries with proxies so that they're more easily
+    // upgradable. All contracts that link libraries should be upgraded to instead link the proxied
+    // library.
+    // To ensure this actually happens, we check that all contracts that link libraries are marked
+    // as needing to be redeployed.
+    // TODO(asa): Remove this check after release 1.
+    const linksLibrariesWithoutChanges = contracts.map(
+      (c) =>
+        dependencies.get(c).length &&
+        !Object.keys(report.contracts).includes(c) &&
+        !c.endsWith('Test')
+    )
+    if (linksLibrariesWithoutChanges.some((b) => b)) {
+      contracts.forEach((c, i) => {
+        if (linksLibrariesWithoutChanges[i]) {
+          console.log(
+            `${c} links ${dependencies.get(c)} and needs to be upgraded to link proxied libraries.`
+          )
+        }
+      })
+      throw new Error('All contracts linking libraries should be upgraded in release 1')
+    }
+
+    const release = async (contractName: string) => {
```

Comment:
This is a pretty massive function, I would recommend splitting it up for readability purposes